database/email.go

```diff
@@ -33,6 +55,55 @@ func (db *DB) EmailCreate(ctx context.Context, m EmailMessage) error {
 	return nil
 }
 
+// EmailLockAndFetch locks up to batchSize records with the given lockId and
+// returns up to batchSize locked entries. Some of the returned entries might
+// not have been locked during the current execution.
+func (db *DB) EmailLockAndFetch(ctx context.Context, lockID string, batchSize int64) (msgs []EmailMessage, err error) {
+	// Find out how many entries are already locked by this id. Maybe we don't
+	// need to lock any additional ones.
+	filter := bson.M{
+		"locked_by":       lockID,
+		"failed_attempts": bson.M{"$lt": EmailMaxSendAttempts},
+		"sent_at":         nil,
+	}
+	count, err := db.staticEmails.CountDocuments(ctx, filter)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to count locked email messages")
+	}
+	// Lock some more entries in order to fill the batch.
+	// We select entries that haven't failed more times than the limit, aren't
+	// sent yet,
+	filterLock := bson.M{
+		"failed_attempts": bson.M{"$lt": EmailMaxSendAttempts},
+		"sent_at":         nil,
+		"$or": []interface{}{
+			bson.M{"locked_by": ""},
+			bson.M{"locked_at": bson.M{"$lt": time.Now().UTC().Add(-emailLockTTL)}},
+		},
+	}
+	updateLock := bson.M{"$set": bson.M{
+		"locked_by": lockID,
+		"locked_at": time.Now().UTC(),
+	}}
+	for i := int64(0); i < batchSize-count; i++ {
+		sr := db.staticEmails.FindOneAndUpdate(ctx, filterLock, updateLock)
```

Comment:
This was my previous iteration on this and it worked worse in a situation where you have multiple servers trying to lock messages at the same time.

The current approach ensures the server will be able to fill the batch (provided there are enough messages to send) and send a full batch, while the approach you suggest would first need to select all messages which were locked by this server (because some other server might have overwritten the locks in a data race) and then attempt to lock new ones if it wants to fill up the batch.